### ISSUE TITLE
Documentation:  Add a section to the context page which describes important details of how the context is evaluated.

### ISSUE THREAD

**Issue by gamagan** [typst/typst#5538]
### Description

# Request 

The `context` page of the docs site could benefit from an additional section, which details low-level behavior of how the `context` object operates. 
https://typst.app/docs/reference/context/

State management in Typst is very different from how it is done in most other software packages. The implementation makes sense in the context of Typst, but it can be challenging to understand for users not accustomed to it.

The Typst `context` is a concept that is technical, and it is technical enought that, in generaly, it's not possible to just explain it at a high level. Because of the nature of state management in Typst, low-level details cannot be escaped. Instead of hiding them, they should be made clear, so that it can bring clarity to how Typst processes the data related to contextual updates. This way, using the context will seem less magical.

To help with the documentation process, below is some info from the Typst discord, as described by user `PgSuper`.

# Info

The way counters work is that it does basically a `query()` of all counter `.step()` and `.update()` until the location where you run `counter.get()`, adding 1 for each `step()` and updating accordingly, returning the calculated value.

But, what is the "location" where you run `counter.get()`? A location is always tied to an element in the document, e.g. when you place a heading, you can refer to where that heading appears in the document.

A location is not something like x=29345, y=203  -- that's an important distinction. A location is like a position in the compiler's element tree, so to speak.

A `context { }` section creates an element which must be placed in the document so that it gains a location. It then provides that location to `.get()` so it knows:  query all steps until that location.

# Example Code

Consider 2 similar function definitions, where the idea is the following: 
Every time the `GetCounter()` function is called, it first updates a global `counter` variable, and then returns the current value of that variable, concatenated to a prefix string and suffix string. 

Thus, successive calls to `GetCounter()` should return: 
        (-1-), (-2-), ..., (-n-)

```typst
#let gCounter = counter("my-counter")

// The context is outside of the function's body, so the ".step()" is a child of the generated context element.
// This function does not achieve its goal.
#let GetCounter1() = context {
    gCounter.step()
    let num = gCounter.get().last()
    return "(-" + str(num) + "-)"
}

// The counter's ".step()" call is outside of the context.
// This is the correct way to implement the GetCounter() functionality described above.
#let GetCounter2() = {
    gCounter.step()

    // note: no return
    context {
        let num = gCounter.get().last()
        [ (-  #num - ) ]
    }
}

```

The first function is conceptually similar to this:

```typst
// Notice that the context provided by the callback doesn't change inside its body:
#let GetCounter1() = [#context-provider(ctx => {
  gCounter.step()    // basically updates a different context than 'ctx'
  let num = gCounter.get(ctx).last()
  return "(-" + str(num) + "-)"
})]
```

The first function, `GetCounter1`, has 2 deficiencies: 
1. The context used when the `gCounter.step()` is called is not the same as the encompasing `context {...`.
2. It is using `return`, so it discards the marker element returned by `.step()`. A counter step only takes effect if you place what it returns in the document.

The second function works correctly.

The main difference between the 2 functions is that the placement of the `context` keyword changes the "location" where the context is evaluated in the element tree. When it is outside, as in `GetCounter1()`, the location of the context evaluation is "before" the `.step()` call, since the `context` element is its parent in the element tree. Conversely,  when the `context` is inside the function, it's "after" the `.step()` call, and thus the step update happens before the context evaluation.

So, when the `context` is inside the `GetCounter()` function,  `.get()`'s query for everything "before" it will include that `.step()` call -- when the `context` is outside, it won't.
In other words, if you add a bunch of `.step()` calls inside a context block, a `.get()` afterwards (inside the same context block) will return the same value as if the `.step()` weren't there, precisely because:

**The context element always gets information from before the block started.**

**Comment by gamagan** [comment]
TO OTHER USERS READING THIS: 

If you have additional information regarding the `context` and `counter` elements, or about state management in general, please add it to this post. It could also be info that you feel is pertinent, or which might have tripped you up while working with these objects. This will give more data to the documentation team to work with.
